### PR TITLE
Spiked shields do something, add some spiked armor

### PR DESCRIPTION
**Pull request by chaosvolt**
#### Summary

SUMMARY: Features "Spiked shields can now damage attackers, add spiked leather jackets and vests, fix a bug with shields"

#### Purpose of change

As promised, I said I'd eventually add support for spiked shields actually affecting melee attacks eventually.

Along the way I also discovered a bug to fix: shields don't trigger ranged blocks on hitting the arms if you're wielding the shield, because the code only checks if the shield covers that part, not if it's specifically WORN on that part.

Also misc, fixes large spiked shields not having the right block tec.

#### Describe the solution

C++ changes:
1. In character.cpp, updated `Character::on_hit` so that hits also check if the body part hit had an item with the `SPIKED_ARMOR` flag on it, that also has stabbing damage. If so, add a chance based off the item's coverage percent (75% of coverage) to jab the attacker for the amount of stabbing damage the item deals in melee.
2. In melee.cpp, updated `Character::block_hit` so that a successful block, with an item that has the `SPIKED_BLOCK` flag, will jab the attack with the same odds and damage. Done such that a spiked shield can both trigger this while wielded, but also won't trigger it twice over on hitting your arm. Since the item in question may potentially not be a wearable item (if we add weapons with the flag), if no coverage is found it sets a default result. 
3. Also in character.cpp, updated `is_covered_by_shield` so that it allows shield items not currently worn to ignore if it covers the affected body part, to fix the case of ranged blocks for attacks hitting a part normally covered by a wielded shield.

JSON changes:
1. Defined `SPIKED` flag in flags.json.
2. Added `SPIKED` flag to spiked shields.
3. Added `SPIKED` flag to pickelhaube.
4. Added two new items: spiked leather jacket and spiked leather vest. Both are modifications of the armored leather jacket/vest that add the same damage as the spike item and gain the `SPIKED` flag.
5. Added recipes for the relevant items, a level higher with the street fighting book as an early booklearn.
6. Added spiked leather jackets and vests to a few relevant NPC gear lists, in particular defining `NC_THUG_coat`.
7. Set large spiked shield to have `WBLOCK_3`, as was spotted in the BN discord.

#### Describe alternatives you've considered

1. Adding `SPIKED` to some weapons that might logically impale enemies when used to block? Not sure.
2. Added spiked leather gloves, probably too small to really justify it I feel.
3. Whenever I get around to adding armored leather chaps and pants, I'll probably add spiked versions too.

#### Testing

1. Checked affected JSON for syntax and lint errors.
2. Compiled and load-tested.
3. Tested that wielding a riot shield now properly interacted with ranged attacks hitting the left and right arm.
4. Wore the shield again to confirm it still also affects attacks as normal.
5. Swapped in a spiked shield and spawned a zombie.
6. Confirmed that blocking with it would inflict damage, whether worn or wielded, even when it hits the arms.
8. Let it commit suicide on the spikes, confirmed this doesn't do anything weird.
9. Testing wearing a spiked leather jacket, confirmed zombies could hurt themselves on it.
10. Debugged my way to a bandit camp, confirmed the thus there had jackets and such present in the itemgroup defined for them, including spiked jackets and vests.
11. Checked affected C++ files for astyle.

#### Additional context

I've now discovered that I clearly need to make a project out of adding a lot more itemgroups to flesh out and flavor-ify NPC gear lists.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Meh design, since it's a copypaste from thorns. Has some problems:
* Damage tied to damage when you swing the item as weapon
* Only covers stabbing damage
* Chance to hit back tied to coverage and not customizable (but see next point)
* Spiked hit back not tied to armor actually blocking the hit
* Dodging the attack means you don't hit back

You could improve the first three by giving it a statblock inside a structure, like you have for the itype components (gun, ammo, wearable, comestible etc.). The statblock could hold the damage instance and hit-back chance.
Last two are hard to fix, so I certainly won't require them being addressed, but if you have ideas, it may be worth considering something more complex here.

**Comment by chaosvolt** [comment]
Hmm. Fair points, not 100% sure how easy it'd be to handle the first three points. Might close for now and re-visit another time if I get any ideas, it's been a busy week and next week is looking to be hectic too, plus this is a fairly minor feature as it is.

File: src/melee.cpp [hunk 1]
```diff
@@ -1801,6 +1801,36 @@ bool Character::block_hit( Creature *source, bodypart_id &bp_hit, damage_instanc
...
+            } else if( g->u.sees( *this ) ) {
+                add_msg( _( "%1$s's %2$s impales %3$s in mid-attack!" ),
+                         name, shield.display_name(),
+                         source->disp_name() );
+            }
+            damage_instance spike_damage;
+            spike_damage.add_damage( DT_STAB, spike );
+            // In general, critters don't have separate limbs
+            // so safer to target the torso
+            source->deal_damage( this, bodypart_id( "torso" ), spike_damage );
```
Comment (scarf005) [inline src/melee.cpp:1829]: - prefer `get_player_character` over `g->u` (it's a modern alias)
- maybe duplicate logics should be moved to `melee_spike.cpp` for ease of maintenance?

**Review by scarf005 — CHANGES_REQUESTED** [review]

**Comment by chaosvolt** [comment]
Oh right, I'd forgotten about this. As mentioned, been busy a lot so oughta close this for the time being, and possibly work on it if I get the time and energy to figure out the problems Coolthulhu brought up previously. Sorry about that. ^^"

**Pull request closed without merging**